server/routerlicious/packages/routerlicious-base/src/alfred/runner.ts

```diff
@@ -104,43 +105,47 @@ export class AlfredRunner implements IRunner {
 		);
 		const isSignalUsageCountingEnabled = this.config.get("usage:signalUsageCountingEnabled");
 
-		// Register all the socket.io stuff
-		configureWebSocketServices(
-			this.server.webSocketServer,
-			this.orderManager,
-			this.tenantManager,
-			this.storage,
-			this.clientManager,
-			createMetricClient(this.metricClientConfig),
-			winston,
-			maxNumberOfClientsPerDocument,
-			numberOfMessagesPerTrace,
-			maxTokenLifetimeSec,
-			isTokenExpiryEnabled,
-			isClientConnectivityCountingEnabled,
-			isSignalUsageCountingEnabled,
-			this.redisCache,
-			this.socketConnectTenantThrottler,
-			this.socketConnectClusterThrottler,
-			this.socketSubmitOpThrottler,
-			this.socketSubmitSignalThrottler,
-			this.throttleAndUsageStorageManager,
-			this.verifyMaxMessageSize,
-			this.socketTracker,
-			this.revokedTokenChecker,
-		);
-
-		// Listen on provided port, on all network interfaces.
-		httpServer.listen(this.port);
 		httpServer.on("error", (error) => this.onError(error));
 		httpServer.on("listening", () => this.onListening());
 
-		// Start token manager
-		if (this.tokenRevocationManager) {
-			this.tokenRevocationManager.start().catch((error) => {
-				// Prevent service crash if token revocation manager fails to start
-				Lumberjack.error("Failed to start token revocation manager.", undefined, error);
-			});
+		if (cluster.isPrimary && this.server.webSocketServer === null) {
+			// Listen on provided port, on all network interfaces.
+			httpServer.listen(this.port);
+		} else {
+			// Register all the socket.io stuff
+			configureWebSocketServices(
+				this.server.webSocketServer,
+				this.orderManager,
+				this.tenantManager,
+				this.storage,
+				this.clientManager,
+				createMetricClient(this.metricClientConfig),
+				winston,
+				maxNumberOfClientsPerDocument,
+				numberOfMessagesPerTrace,
+				maxTokenLifetimeSec,
+				isTokenExpiryEnabled,
+				isClientConnectivityCountingEnabled,
+				isSignalUsageCountingEnabled,
+				this.redisCache,
+				this.socketConnectTenantThrottler,
+				this.socketConnectClusterThrottler,
+				this.socketSubmitOpThrottler,
+				this.socketSubmitSignalThrottler,
+				this.throttleAndUsageStorageManager,
+				this.verifyMaxMessageSize,
+				this.socketTracker,
+				this.revokedTokenChecker,
+			);
+			// Listen on primary thread port, on all network interfaces.
+			httpServer.listen(cluster.isPrimary ? this.port : 0);
```

Comment:
`cluster` is the [Node.js cluster module](https://nodejs.org/docs/latest-v18.x/api/cluster.html#workerdisconnect). `cluster.isPrimary` tells you if this is the main thread or a worker thread. The main thread is used for managing worker threads but does not handle traffic itself; instead it forwards requests on to workers

Reply:
To quote the cluster module docs:
> `server.listen(0)` Normally, this will cause servers to listen on a random port. However, in a cluster, each worker will receive the same "random" port each time they do listen(0). In essence, the port is random the first time, but predictable thereafter. To listen on a unique port, generate a port number based on the cluster worker ID.
So primary listens on the port we define (i.e. 3000), then all the child workers listen on a random, different port. The primary forwards all connections to those workers.

Reply:
Neat! Thanks for the explanation!